src/IdentityDecorator.php

```diff
@@ -0,0 +1,135 @@
+<?php
+/**
+ * CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Authorization;
+
+use ArrayAccess;
+use InvalidArgumentException;
+
+/**
+ * An decorator implementing the IdentityInterface.
+ *
+ * This decorator is intended to wrap the application defined identity
+ * object and proxy attributes/methods to and 'mixin' the can() method.
+ *
+ * The decorated identity must implement ArrayAccess or be an array.
+ */
+class IdentityDecorator implements IdentityInterface
+{
+    /**
+     * Identity data
+     *
+     * @var array|\ArrayAccess
+     */
+    protected $identity;
+
+    /**
+     * Authorization Service
+     *
+     * @var \Authorization\AuthorizationServiceInterface
+     */
+    protected $authorization;
+
+    /**
+     * Constructor
+     *
+     * @param \Authorization\AuthorizationServiceInterface $service The authorization service.
+     * @param array|\ArrayAccess $identity Identity data
+     * @throws InvalidArgumentException When invalid identity data is passed.
+     */
+    public function __construct(AuthorizationServiceInterface $service, $identity)
+    {
+        if (!is_array($identity) && !$identity instanceof ArrayAccess) {
+            $type = is_object($identity) ? get_class($identity) : gettype($identity);
+            $message = sprintf('Identity data must be an `array` or implement `ArrayAccess` interface, `%s` given.', $type);
+            throw new InvalidArgumentException($message);
+        }
+
+        $this->authorization = $service;
+        $this->identity = $identity;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function can($action, $resource)
+    {
+        return $this->authorization->can($this, $action, $resource);
```

Comment:
Should `$this` be passed to the service, or maybe `$this->identity`?
Policies may typehint some internal identity object or do `instance_of` checks.

Reply:
That is a good point. I was concerned that passing the wrapped identity would make combining permissions/policy checks defined in other resources harder. For example passing the unwrapped user would make something like the following not possible.

```php
function update($user, $thing)
{
  return $user->can('update', $this->category);
}
```

Reply:
We can have some kind of `getOriginal` method in the identity. This way we could combine checks in policies and still access the original object for some internal checks.

For example:
```php
function canUpdate($identity, $object) {
   return $identity->can(‘access’, $object) || $identity->getOriginal() instanceof Admin);
}

Reply:
I can add a getOriginal() method to the interface and decorator.